src/hotspot/cpu/s390/templateInterpreterGenerator_s390.cpp

```diff
@@ -1134,7 +1134,15 @@ void TemplateInterpreterGenerator::generate_fixed_frame(bool native_call) {
   __ z_agr(Z_locals, Z_esp);
   // z_ijava_state->locals - i*BytesPerWord points to i-th Java local (i starts at 0)
   // z_ijava_state->locals = Z_esp + parameter_count bytes
-  __ z_stg(Z_locals, _z_ijava_state_neg(locals), fp);
+
+  __ z_lgr(Z_R0, Z_R1); // preserve Z_R1, holding cache offset
+
+  __ z_sgrk(Z_R1, Z_locals, fp); // Z_R1 = Z_locals - fp();
+  __ z_srlg(Z_R1, Z_R1, Interpreter::logStackElementSize);
+  // Store relativized Z_locals, see frame::interpreter_frame_locals().
+  __ z_stg(Z_R1, _z_ijava_state_neg(locals), fp);
+
+  __ z_lgr(Z_R1, Z_R0); // restore R1
```

Comment:
Why don't you use Z_R0 for the relativation calculations and leave Z_R1 untouched? That will avoid the save/restore overhead.